File: api-chi/cmd/services/movie_category.go
package services

import (
	"api-chi/cmd/config"
	"api-chi/cmd/models"

	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"
)

type MovieCategoryService struct {
	Conn *pgxpool.Pool
}

func (s *MovieCategoryService) Open() error {
	config.LoadConfig()
	conn, err := pgxpool.New(config.CTX, config.POSTGRES_URL)
	if err != nil {
		return err
	}

	// Assign connection
	s.Conn = conn
	return nil
}

func (s *MovieCategoryService) Close() {
	s.Conn.Close()
}

func (s *MovieCategoryService) Create(input *models.MovieCategory) (string, error) {
	// Execute SQL
	sql := "SELECT * FROM create_movie_category(@name);"
	args := pgx.NamedArgs{
		"name": input.Name,
	}
	value := ""
	err := s.Conn.QueryRow(config.CTX, sql, args).Scan(&value)
	if err != nil {
		return value, err
	}

	// If success return nil
	return value, nil
}

func (s *MovieCategoryService) Update(input *models.MovieCategory) (models.MovieCategory, error) {
	// Execute SQL
	sql := "SELECT * FROM update_movie_category(@id, @name);"
	args := pgx.NamedArgs{
		"id":   input.Id,
		"name": input.Name,
	}
	value := models.MovieCategory{}
	err := s.Conn.QueryRow(config.CTX, sql, args).Scan(&value.Id, &value.Name)
	if err != nil {
		return value, err
	}

	// If success return nil
	return value, nil
}

func (s *MovieCategoryService) Remove(id *string) (string, error) {
	// Execute SQL
	sql := "SELECT * FROM remove_movie_category(@id);"
	args := pgx.NamedArgs{
		"id": *id,
	}
	value := ""
	err := s.Conn.QueryRow(config.CTX, sql, args).Scan(&value)
	if err != nil {
		return value, err
	}

	// If success return nil
	return value, nil
}

func (s *MovieCategoryService) Count() (int, error) {
	// Execute SQL
	sql := "SELECT * FROM count_movie_category();"
	value := 0
	err := s.Conn.QueryRow(config.CTX, sql).Scan(&value)
	if err != nil {
		return value, err
	}

	// If success return nil
	return value, nil
}

func (s *MovieCategoryService) GetAll(search *string, limit *int, page *int) ([]models.MovieCategory, error) {
	// Set default range for limit
	if *limit < 10 {
		*limit = 10
	} else if *limit > 50 {
		*limit = 50
	}

	// Set default range for page
	if *page < 1 {
		*page = 0
	} else {
		*page = *page - 1
	}

	// Execute SQL
	sql := "SELECT * FROM get_all_movie_category(@search, @limit, @page);"
	args := pgx.NamedArgs{
		"search": *search,
		"limit":  *limit,
		"page":   *page,
	}
	value := []models.MovieCategory{}
	rows, err := s.Conn.Query(config.CTX, sql, args)
	if err != nil {
		return value, err
	}
	for rows.Next() {
		item := models.MovieCategory{}

		if err := rows.Scan(&item.Id, &item.Name); err != nil {
			return nil, err
		}

		value = append(value, item)
	}

	// If success return nil
	return value, nil
}

File: api-fastify/src/api/services/movie_category.ts
import { EnumMessage, ReturnMessage } from "@/types/message"
import pool from "@/libs/config/database_postgres"
import { MovieCategory } from "@/types/movie_category"

export default class MovieCategoryService {
  public async Create(input: MovieCategory): Promise<ReturnMessage> {
    return pool
      .withClient(async (client) => {
        const result = await client.query(
          "SELECT * FROM create_movie_category($1)",
          [input.name],
        )
        const id: string = result.rows[0].create_movie_category
        return { success: true, message: EnumMessage.CREATE_SUCCESS, data: id }
      })
      .catch((error) => this.handleError(error, EnumMessage.CREATE_FAILED))
  }

  public async Update(input: MovieCategory): Promise<ReturnMessage> {
    return pool
      .withClient(async (client) => {
        const result = await client.query(
          "SELECT * FROM update_movie_category($1, $2)",
          [input.id, input.name],
        )

        if (!Array.isArray(result.rows) || result.rows.length === 0)
          throw new Error("Update failed, no movie category returned.")

        const row = result.rows[0]
        const movie_category: MovieCategory = {
          id: row.id,
          name: row.name,
        }

        return {
          success: true,
          message: EnumMessage.UPDATE_SUCCESS,
          data: movie_category,
        }
      })
      .catch((error) => this.handleError(error, EnumMessage.UPDATE_FAILED))
  }

  public async Remove(id: string): Promise<ReturnMessage> {
    return pool
      .withClient(async (client) => {
        const result = await client.query(
          "SELECT * FROM remove_movie_category($1)",
          [id],
        )
        const removeId: string = result.rows[0].remove_movie_category
        return {
          success: true,
          message: EnumMessage.REMOVE_SUCCESS,
          data: removeId,
        }
      })
      .catch((error) => this.handleError(error, EnumMessage.REMOVE_FAILED))
  }

  public async Count(): Promise<ReturnMessage> {
    return pool
      .withClient(async (client) => {
        const result = await client.query(
          "SELECT * FROM count_movie_category()",
        )
        const data: number = parseInt(result.rows[0].count_movie_category)
        return { success: true, message: EnumMessage.GET_SUCCESS, data: data }
      })
      .catch((error) => this.handleError(error, EnumMessage.GET_FAILED))
  }

  public async GetAll(
    search: string,
    limit: number,
    page: number,
  ): Promise<ReturnMessage> {
    return pool
      .withClient(async (client) => {
        const result = await client.query(
          "SELECT * FROM get_all_movie_category($1, $2, $3)",
          [search, limit, page - 1],
        )

        let data: MovieCategory[] = Array.isArray(result.rows)
          ? result.rows.map((row) => {
              return {
                id: row.id,
                name: row.name,
              } as MovieCategory
            })
          : []

        return { success: true, message: EnumMessage.GET_SUCCESS, data: data }
      })
      .catch((error) => this.handleError(error, EnumMessage.GET_FAILED))
  }

  private handleError(error: any, message: EnumMessage): ReturnMessage {
    console.error(error)
    return {
      success: false,
      message,
      data: undefined,
      error_detail: error.message,
    }
  }
}
